src/client/handlers/VaultsSecretsMkdir.ts

```diff
@@ -2,50 +2,77 @@ import type { DB } from '@matrixai/db';
 > {
-  public handle = async (
-    input: ClientRPCRequestParams<SecretMkdirMessage>,
-  ): Promise<ClientRPCResponseResult<SuccessMessage>> => {
-    const { vaultManager, db } = this.container;
-    await db.withTransactionF(async (tran) => {
-      const vaultIdFromName = await vaultManager.getVaultId(
-        input.nameOrId,
-        tran,
-      );
-      const vaultId =
-        vaultIdFromName ?? vaultsUtils.decodeVaultId(input.nameOrId);
-      if (vaultId == null) {
-        throw new vaultsErrors.ErrorVaultsVaultUndefined();
+  public handle = async function* (
+    input: AsyncIterable<ClientRPCRequestParams<SecretDirMessage>>,
+  ): AsyncGenerator<ClientRPCResponseResult<SuccessWithErrorMessage>> {
+    const { vaultManager, db }: { vaultManager: VaultManager; db: DB } =
+      this.container;
+    // Create a record of secrets to be removed, grouped by vault names
+    const vaultGroups: Record<string, Array<string>> = {};
+    const dirPaths: Array<[string, string]> = [];
+    let metadata: any = undefined;
+    for await (const secretDirMessage of input) {
+      if (metadata == null) metadata = secretDirMessage.metadata ?? {};
+      dirPaths.push([secretDirMessage.nameOrId, secretDirMessage.dirName]);
+    }
+    dirPaths.forEach(([vaultName, dirPath]) => {
+      if (vaultGroups[vaultName] == null) {
+        vaultGroups[vaultName] = [];
       }
-      await vaultManager.withVaults(
-        [vaultId],
-        async (vault) => {
-          await vaultOps.mkdir(vault, input.dirName, {
-            recursive: input.recursive,
-          });
-        },
-        tran,
-      );
+      vaultGroups[vaultName].push(dirPath);
     });
-    return {
-      success: true,
-    };
+    // Use the grouping to create directories for each vault in one commit
+    yield* db.withTransactionG(
+      async function* (tran): AsyncGenerator<SuccessWithErrorMessage> {
+        for (const [vaultName, dirPaths] of Object.entries(vaultGroups)) {
+          const vaultIdFromName = await vaultManager.getVaultId(
+            vaultName,
+            tran,
+          );
+          const vaultId =
+            vaultIdFromName ?? vaultsUtils.decodeVaultId(vaultName);
+          if (vaultId == null) {
+            throw new vaultsErrors.ErrorVaultsVaultUndefined();
+          }
+          yield* vaultManager.withVaultsG(
+            [vaultId],
+            async function* (vault): AsyncGenerator<SuccessWithErrorMessage> {
+              const response = vaultOps.makeDirectories(vault, dirPaths, {
+                recursive: metadata?.options?.recursive,
+              });
+              for await (const data of response) {
+                if (data.success) yield { success: true };
+                if (data.error != null) {
+                  yield {
+                    success: false,
+                    error: { code: data.error.code, path: data.error.path },
+                  };
+                }
```

Comment:
This current approach that I have highlighted actually either gets `read 0` or `ErrorRPCParse`. This is probably not even the right solution here, so I haven't put in much effort in getting this prototype error handler working.